### ISSUE TITLE
KMP_DUPLICATE_LIB_OK=TRUE with a second libomp in the process causes intermittent autograd deadlock and MPS sync hangs (macOS)

### ISSUE THREAD

**Issue by KalenJosifovski** [pytorch/pytorch#193520]
## Summary

On macOS, when a process loads **two OpenMP runtimes** (conda-forge `llvm-openmp` via a conda package such as scikit-learn, plus torch's bundled `libomp.dylib`), training **intermittently deadlocks** under the documented `KMP_DUPLICATE_LIB_OK=TRUE` workaround instead of failing loudly.

## Environment

- torch 2.10.0 (pip wheel), macOS arm64 (M1 Pro)
- Process also loads conda-forge `llvm-openmp` 22.x (pulled via scikit-learn's `_openmp_mutex` dependency)

## Expected behavior

Either work correctly, or fail loudly and deterministically. The current behavior is a nondeterministic hang at 0% CPU — identical runs sometimes complete and sometimes hang forever, which is extremely expensive to diagnose.

## Actual behavior

- Without the suppression: `import torch` hard-aborts with `OMP: Error #15` (multiple copies of the OpenMP runtime).
- With `KMP_DUPLICATE_LIB_OK=TRUE`: training **intermittently hangs**. Stack samples show the main thread parked in the OpenMP fork/barrier path:

```
__kmpc_fork_call -> __kmp_fork_call -> __kmp_hyper_barrier_gather -> __psynch_cvwait
```

with OpenMP worker threads idle on condition variables — while `ps` shows the process "running" at ~0% CPU and no progress.

- On MPS the same root manifests as a device-sync hang:

```
at::mps::MPSStream::copy_and_sync -> at::mps::MPSStream::synchronize
```

i.e. autograd blocked on the OpenMP barrier while holding the MPS sync.

## Hypothesis

With two runtime copies, fork/barrier coordination is inconsistent; `KMP_DUPLICATE_LIB_OK=TRUE` converts the loud failure into a silent race. The race only triggers reliably with sufficiently large autograd graphs (a small model in the same environment ran fine for weeks).

## Repro sketch

- `pip install torch` + a conda package linking `llvm-openmp` (e.g. `scikit-learn`) in one environment on macOS.
- Train a moderately sized model (e.g. a 2048-dim MPNN with a ~1.6k-output head) for several epochs; expect intermittent hangs with the main thread in `__kmp_fork_call` / barrier wait.

## Ask

1. Make the coexistence fail loudly/detectably (or fix the race).
2. Document the `KMP_DUPLICATE_LIB_OK` interaction — the suppression is currently documented as "unsafe, unsupported" but the realistic outcome (intermittent deadlock) is worse than the abort it avoids.
3. Long-standing: consider not bundling `libomp` (allow using the process OpenMP runtime).
4. Note the MPS interplay — an MPS sync hang can be the *symptom* of this OpenMP issue, worth a diagnostic hint.

cc @malfet @aditvenk @kulinseth @DenisVieriu97 @jhavukainen @Isalia20
